File: Sources/App/DangerPattern.swift
import Foundation

/// 授权模式：决定 agent 发起/执行的操作在多大程度上需要用户确认。
///
/// - `.safe`（默认）：只拦截命中危险模式的操作，其余自动放行。
/// - `.askAll`：每一条命令都要确认（供极度谨慎的用户手动升级）。
/// - `.auto`：全程免确认，等同 Claude Code 的 `--dangerously-skip-permissions`。
///
/// 默认落在 `.safe`，是「默认谨慎 + 渐进放权」策略的锚点：
/// 首装用户对"手机遥控 Mac 执行命令"的风险尚未建立认知，必须从最安全一档起步。
enum ApprovalMode: String, Codable, CaseIterable {
    case safe
    case askAll = "askAll"
    case auto
}

/// 一次危险命中：`code` 是稳定的机器可读标识（iOS 端据此本地化文案），
/// `detail` 是命中的原始片段，供用户看到具体命令上下文。
struct DangerHit: Codable, Equatable {
    var code: String
    var detail: String
}

/// 危险命令检测器：本地内置的模式匹配，**不信任 agent 自报**。
///
/// 检测点是「命令进入 agent 之前」—— 用户在 iPhone 上发的消息正文。
/// 受限于 opencode 是黑盒 TUI（内部调用哪些工具我们看不到），
/// 只能拦"入口"，无法拦"agent 在会话中途自行推断出的 shell 命令"；
/// 这已经是当前架构下能做到的最诚实的 L3 简化版。
enum DangerPattern {
    /// 危险规则表。`code` 保持稳定，改动只加不减（避免已持久化的 alwaysAllow 失效）。
    /// `pattern` 是正则，`extract` 从原文里摘一段人类可读的命中片段。
    private static let rules: [(code: String, pattern: String, hint: String)] = [
        // 最高危：删除根 / 家目录
        ("rm_root", #"rm\s+(-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*|-rf|-fr|--recursive\s+--force)\s+/(\s|$)"#, "rm -rf /"),
        // 递归删除
        ("recursive_delete", #"rm\s+(-rf|-fr|--recursive|-R)\b"#, "recursive delete"),
        // 强制推送（--force-with-lease 是安全变体，单独排除）
        ("force_push", #"git\s+push\b[^\n]*--force(?!-with-lease)"#, "git push --force"),
        // 硬重置
        ("git_reset_hard", #"git\s+reset\s+--hard\b"#, "git reset --hard"),
        // 管道执行远程脚本
        ("pipe_to_shell", #"(curl|wget)\b[^\n]*\|\s*(sudo\s+)?(ba|z|fi)?sh\b"#, "curl | sh"),
        // 提权
        ("sudo", #"\bsudo\b"#, "sudo"),
        // 开放权限
        ("chmod_777", #"chmod\s+(-R\s+)?0?777\b"#, "chmod 777"),
        // 写设备
        ("write_device", #">\s*/dev/"#, "write to /dev/"),
        // 磁盘镜像写入
        ("dd_disk", #"\bdd\s+if="#, "dd if="),
        // 破坏性数据库操作
        ("drop_table", #"\bdrop\s+table\b"#, "DROP TABLE"),
        ("truncate_table", #"\btruncate\s+table\b"#, "TRUNCATE TABLE"),
    ]

    /// 检测文本里命中的危险操作，返回命中的 `DangerHit` 列表（可能为空）。
    static func detect(in text: String) -> [DangerHit] {
        var hits: [DangerHit] = []
        for rule in rules {
            if let regex = try? NSRegularExpression(pattern: rule.pattern, options: [.caseInsensitive]),
               regex.firstMatch(in: text, options: [], range: NSRange(text.startIndex..., in: text)) != nil {
                hits.append(DangerHit(code: rule.code, detail: rule.hint))
            }
        }
        return hits
    }

    /// 所有已知的 danger code，供「总是允许」清单做校验（拒绝未知 code 注入）。
    static var knownCodes: Set<String> {
        Set(rules.map { $0.code })
    }
}

File: Sources/BrewPingwinDesktop/src-tauri/src/services/danger_pattern.rs
//! 危险操作检测：本地内置的模式匹配，**不信任 agent 自报**。
//!
//! 与 macOS `Sources/App/DangerPattern.swift` 一比一对应：
//! - `code` 必须保持完全一致（iOS 端按 `danger.<code>` 本地化文案）；
//! - `detail` 是命中的原始片段提示，供用户看到具体命令上下文；
//! - 规则表只加不减，避免已持久化的 alwaysAllow 白名单失效。
//!
//! 检测点是「命令进入 agent 之前」—— 用户在 iPhone / Watch 上发出的消息正文。
//! 受限于各 Agent 都是黑盒 TUI（内部调用哪些工具我们看不到），
//! 只能拦"入口"，无法拦"agent 在会话中途自行推断出的 shell 命令"；
//! 这已经是当前架构下能做到的最诚实的 L3 简化版。

use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::sync::OnceLock;

/// 一次危险命中。`code` 是稳定的机器可读标识，`detail` 供人阅读。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DangerHit {
    pub code: String,
    pub detail: String,
}

/// 单条危险规则。
struct Rule {
    /// 稳定的机器可读标识（与 macOS 端逐字相同）。
    code: &'static str,
    /// 正则源码。
    ///
    /// NOTE: 这里用的是 Rust `regex` crate，**不支持 lookaround**。
    /// macOS 端 `force_push` 用了 `(?!-with-lease)` 负向前瞻，
    /// 此处的等价写法是把前瞻改写成显式字符类（见下），语义保持一致：
    /// `--force` 后面只要不是 `-`（即不是 `-with-lease`）就算命中。
    pattern: &'static str,
    /// 命中时展示给用户的短标签。
    hint: &'static str,
}

/// 危险规则表。`code` 保持稳定，改动只加不减。
const RULES: &[Rule] = &[
    // 最高危：删除根 / 家目录
    Rule {
        code: "rm_root",
        pattern: r"rm\s+(-[a-zA-Z]*r[a-zA-Z]*f[a-zA-Z]*|-rf|-fr|--recursive\s+--force)\s+/(\s|$)",
        hint: "rm -rf /",
    },
    // 递归删除
    Rule {
        code: "recursive_delete",
        pattern: r"rm\s+(-rf|-fr|--recursive|-R)\b",
        hint: "recursive delete",
    },
    // 强制推送（--force-with-lease 是安全变体，单独排除）
    Rule {
        code: "force_push",
        pattern: r"git\s+push\b[^\n]*--force(\s|[^-]|$)",
        hint: "git push --force",
    },
    // 硬重置
    Rule {
        code: "git_reset_hard",
        pattern: r"git\s+reset\s+--hard\b",
        hint: "git reset --hard",
    },
    // 管道执行远程脚本
    Rule {
        code: "pipe_to_shell",
        pattern: r"(curl|wget)\b[^\n]*\|\s*(sudo\s+)?(ba|z|fi)?sh\b",
        hint: "curl | sh",
    },
    // 提权
    Rule {
        code: "sudo",
        pattern: r"\bsudo\b",
        hint: "sudo",
    },
    // 开放权限
    Rule {
        code: "chmod_777",
        pattern: r"chmod\s+(-R\s+)?0?777\b",
        hint: "chmod 777",
    },
    // 写设备
    Rule {
        code: "write_device",
        pattern: r">\s*/dev/",
        hint: "write to /dev/",
    },
    // 磁盘镜像写入
    Rule {
        code: "dd_disk",
        pattern: r"\bdd\s+if=",
        hint: "dd if=",
    },
    // 破坏性数据库操作
    Rule {
        code: "drop_table",
        pattern: r"\bdrop\s+table\b",
        hint: "DROP TABLE",
    },
    Rule {
        code: "truncate_table",
        pattern: r"\btruncate\s+table\b",
        hint: "TRUNCATE TABLE",
    },
];

/// 编译后的规则缓存（正则只编译一次）。
fn compiled_rules() -> &'static [(Regex, &'static Rule)] {
    static CACHE: OnceLock<Vec<(Regex, &'static Rule)>> = OnceLock::new();
    CACHE.get_or_init(|| {
        RULES
            .iter()
            .map(|rule| {
                let re = regex::RegexBuilder::new(rule.pattern)
                    .case_insensitive(true)
                    .build()
                    .unwrap_or_else(|e| panic!("DangerPattern 规则 {} 正则非法: {e}", rule.code));
                (re, rule)
            })
            .collect()
    })
}

/// 检测文本里命中的危险操作，按规则表顺序返回（可能为空）。
pub fn detect(text: &str) -> Vec<DangerHit> {
    compiled_rules()
        .iter()
        .filter(|(re, _)| re.is_match(text))
        .map(|(_, rule)| DangerHit {
            code: rule.code.to_string(),
            detail: rule.hint.to_string(),
        })
        .collect()
}

/// 所有已知的 danger code，供「总是允许」清单做校验（拒绝未知 code 注入持久化）。
pub fn known_codes() -> HashSet<String> {
    RULES.iter().map(|r| r.code.to_string()).collect()
}

/// 所有已知 code 的数量（测试与调试用）。
pub fn rule_count() -> usize {
    RULES.len()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn codes(text: &str) -> Vec<String> {
        detect(text).into_iter().map(|h| h.code).collect()
    }

    // TC-DP-01  最高危：删根同时命中 rm_root 与 recursive_delete
    #[test]
    fn rm_root_is_detected() {
        let hits = codes("rm -rf /");
        assert!(hits.contains(&"rm_root".to_string()), "应命中 rm_root: {hits:?}");
        assert!(hits.contains(&"recursive_delete".to_string()));
    }

    // TC-DP-02  递归删除：多种写法都要命中
    #[test]
    fn recursive_delete_variants() {
        for text in [
            "rm -rf ./build",
            "rm -fr /tmp/x",
            "rm --recursive dir",
            "rm -R node_modules",
        ] {
            assert!(
                codes(text).contains(&"recursive_delete".to_string()),
                "{text} 应命中 recursive_delete"
            );
        }
    }

    // TC-DP-03  边界：git push --force 命中，--force-with-lease 不命中
    #[test]
    fn force_push_excludes_force_with_lease() {
        assert!(codes("git push --force").contains(&"force_push".to_string()));
        assert!(codes("git push origin main --force").contains(&"force_push".to_string()));
        assert!(
            !codes("git push --force-with-lease").contains(&"force_push".to_string()),
            "--force-with-lease 是安全变体，不应命中"
        );
        assert!(
            !codes("git push --force-with-lease origin main").contains(&"force_push".to_string())
        );
    }

    // TC-DP-04  git reset --hard
    #[test]
    fn git_reset_hard_is_detected() {
        assert!(codes("git reset --hard HEAD~1").contains(&"git_reset_hard".to_string()));
    }

    // TC-DP-05  管道执行远程脚本
    #[test]
    fn pipe_to_shell_variants() {
        for text in [
            "curl https://example.com/i.sh | sh",
            "wget -qO- https://x/y | bash",
            "curl -s u | sudo bash",
        ] {
            assert!(
                codes(text).contains(&"pipe_to_shell".to_string()),
                "{text} 应命中 pipe_to_shell"
            );
        }
    }

    // TC-DP-06  提权 / 开放权限 / 写设备 / dd
    #[test]
    fn misc_privilege_rules() {
        assert!(codes("sudo apt install x").contains(&"sudo".to_string()));
        assert!(codes("chmod 777 ./data").contains(&"chmod_777".to_string()));
        assert!(codes("chmod -R 0777 ./data").contains(&"chmod_777".to_string()));
        assert!(codes("echo 1 > /dev/sda").contains(&"write_device".to_string()));
        assert!(codes("dd if=/dev/zero of=disk.img").contains(&"dd_disk".to_string()));
    }

    // TC-DP-07  破坏性数据库操作大小写不敏感
    #[test]
    fn database_rules_are_case_insensitive() {
        assert!(codes("DROP TABLE users;").contains(&"drop_table".to_string()));
        assert!(codes("drop table users;").contains(&"drop_table".to_string()));
        assert!(codes("TRUNCATE TABLE logs").contains(&"truncate_table".to_string()));
    }

    // TC-DP-08  边界：普通文本不得误报
    #[test]
    fn benign_text_has_no_hits() {
        for text in [
            "帮我看一下 README",
            "git push origin main",
            "git status",
            "npm run build",
            "rm file.txt",
            "ls -la",
        ] {
            assert!(detect(text).is_empty(), "{text} 不应命中任何规则: {:?}", codes(text));
        }
    }

    // TC-DP-09  契约：code 集合与 macOS 完全一致（只加不减）
    #[test]
    fn known_codes_match_macos_contract() {
        let expected = [
            "rm_root",
            "recursive_delete",
            "force_push",
            "git_reset_hard",
            "pipe_to_shell",
            "sudo",
            "chmod_777",
            "write_device",
            "dd_disk",
            "drop_table",
            "truncate_table",
        ];
        let actual = known_codes();
        assert_eq!(actual.len(), expected.len(), "规则数量与 macOS 不一致");
        for code in expected {
            assert!(actual.contains(code), "缺少 macOS 端已有的 code: {code}");
        }
        // detail 必须非空，否则 iOS 端无法兜底展示
        for hit in detect("sudo rm -rf / && git push --force && drop table t;") {
            assert!(!hit.detail.is_empty(), "{} 的 detail 不应为空", hit.code);
        }
    }

    // TC-DP-10  命中顺序与规则表顺序一致
    #[test]
    fn hits_follow_rule_order() {
        let hits = codes("sudo rm -rf /");
        let rm_root = hits.iter().position(|c| c == "rm_root").unwrap();
        let recursive = hits.iter().position(|c| c == "recursive_delete").unwrap();
        let sudo = hits.iter().position(|c| c == "sudo").unwrap();
        assert!(rm_root < recursive && recursive < sudo, "命中顺序应遵循规则表: {hits:?}");
    }
}
